### PR TITLE
Add maximum validator count

### PR DESCRIPTION
**Pull request by PtitLuca**
## Description

Fixes EDGE-6

This PR adds the following :

- A maximum validator count, allowing only a certain number of validator to be in the set
- Tests for both features

## Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [X] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

## Checklist

- [X] I have assigned this PR to myself
- [X] I have added at least 1 reviewer
- [X] I have tested this code
- [X] I have updated the README and other relevant documents (guides...)
- [X] I have added sufficient documentation both in code, as well as in the READMEs

### CODE REVIEW

**Review by asalem1987 — APPROVED** [review]

**Review by asalem1987 — APPROVED** [review]

**Review by asalem1987 — APPROVED** [review]

**Review by asalem1987 — APPROVED** [review]

File: contracts/staking.sol [hunk 1]
```diff
@@ -116,4 +136,27 @@ contract Staking {
...
+        _lowestValidator = _validators[0];
+
+        for (uint32 i = 1; i < _validators.length; i++) {
+            if (_addressToStakedAmount[_validators[i]] < min) {
+                min = _addressToStakedAmount[_validators[i]];
+                _lowestValidator = _validators[i];
+            }
+        }
+        _minimumStakedAmountByValidator = min;
+    }
```
Comment (Kourin1996) [inline contracts/staking.sol:155]: Scanning all validators every time an account stake or unstake looks not efficient in terms of gas usage.
There are 2 approaches, I think
(1) Contract has MinimumStakeAmountByValidator and its address already. I think you can just compare with this.
(2) You can use some data structure like heap to reduce computation
cc: @zivkovicmilos

File: contracts/staking.sol [hunk 2]
```diff
@@ -68,12 +79,20 @@ contract Staking {
...
+                _appendToValidatorSet(msg.sender);
+            } else {
+                require(
+                    _addressToStakedAmount[msg.sender] > _minimumStakedAmountByValidator,
+                    "Total amount the account staked is less than current minimum staked amount by a validator"
+                );
+
+                _deleteFromValidators(_lowestValidator);
+                _appendToValidatorSet(msg.sender);
+            }
```
Comment (Kourin1996) [inline contracts/staking.sol:92]: Actually _appendToValidatorSet is called in both case of 2nd level if-else block. You can remove else block like this:

```sol
           if (_validators.length >= MaximumNumValidators) {
               
                require(
                    _addressToStakedAmount[msg.sender] > _minimumStakedAmountByValidator,
                    "Total amount the account staked is less than current minimum staked amount by a validator"
                );

                _deleteFromValidators(_lowestValidator);
                _appendToValidatorSet(msg.sender);
            }
           _appendToValidatorSet(msg.sender);
```

**Review by Kourin1996 — CHANGES_REQUESTED** [review]
Basic flow looks ok for me. Let's discuss optimization
@PtitLuca @lazartravica @zivkovicmilos

**Comment by Kourin1996** [comment]
@PtitLuca @lazartravica @zivkovicmilos
I'd like to raise one topic.
How does staking contract decide the validator to be dropped from set if there are multiple candidates in validator set?

**Comment by lazartravica** [comment]
> 

In this implementation we only want to limit the maximum number of validators allowed in the staking smart contract.

If the transaction which does new staking to the smart contract is the one that would exceed this maximum number of validators, it needs to be reverted with no updates and no funds taken from the user.

File: contracts/staking.sol (hunk 2 above)
Comment (PtitLuca) [reply contracts/staking.sol]: Good catch ! Updated in https://github.com/0xPolygon/staking-contracts/pull/10/commits/fb5f135ad666c6e9b41b33992f1ad10e801dd894

File: contracts/staking.sol [hunk 3]
```diff
@@ -8,13 +8,16 @@ contract Staking {
     // Parameters
     uint128 public constant ValidatorThreshold = 1 ether;
     uint32 public constant MinimumRequiredNumValidators = 4;
+    uint32 public constant MaximumNumValidators = 6;
```
[the remark this answers is not in the document]
Comment (PtitLuca) [reply contracts/staking.sol]: 👍 Added in https://github.com/0xPolygon/staking-contracts/pull/10/commits/59687625a809a57facbf7b4b7a3edddc651603d9

**Comment by Kourin1996** [comment]
> > 
> 
> 
> 
> In this implementation we only want to limit the maximum number of validators allowed in the staking smart contract.
> 
> 
> 
> If the transaction which does new staking to the smart contract is the one that would exceed this maximum number of validators, it needs to be reverted with no updates and no funds taken from the user.

@lazartravica Agree with that 👌
@PtitLuca Can you fix implementation based on Lazar's explanation?

File: contracts/Staking.sol [hunk 4]
```diff
@@ -35,7 +36,9 @@ contract Staking {
         _;
     }
 
-    constructor() {}
+    constructor(uint32 maxNumValidators) {
+        _maximumNumValidators = maxNumValidators;
```
Comment (Kourin1996) [inline contracts/Staking.sol:40]: For now, we don't use constructor for pre-deploy. I think you don't need to set _maximumValidators here

**Review by Kourin1996 — APPROVED** [review]
Looks good! I've left small comment. Please check it.

Can you create new PR to update hardcoded bytecode in polygon-edge like this? https://github.com/0xPolygon/polygon-edge/pull/376

File: contracts/Staking.sol (hunk 4 above)
Comment (PtitLuca) [reply contracts/Staking.sol:40]: Thanks ! I'll update that

Comment (PtitLuca) [reply contracts/Staking.sol:40]: As this is the only way to provide a value for testing purposes and considering the fact that the constructor is not used for pre-deploy, we'll leave it for now.

**Review by asalem1987 — APPROVED** [review]

**Review by asalem1987 — APPROVED** [review]

**Review by Aleksao998 — APPROVED** [review]
Great job. Everything works good. i have only 2 suggestions.
1. when we reach limit of max validators transaction fails. But in the log I do not see what was the reason why the transaction failed. Not sure can we add failed reason but it would be great.
Screen Shot 2022-02-03 at 4 38 11 PM
2. Just for consideration, we can open task to add additional info (like max-validators) when printing info (npm run info)

**Review by zivkovicmilos — APPROVED** [review]
Lookin' good 💯

**Review by Kourin1996 — APPROVED** [review]

File: contracts/Staking.sol [hunk 5]
```diff
@@ -35,7 +36,10 @@ contract Staking {
         _;
     }
 
-    constructor() {}
+    constructor(uint32 minNumValidators, uint32 maxNumValidators) {
```
Comment (zivkovicmilos) [inline contracts/Staking.sol:39]: What happens when you deploy this contract without specifying these values, will they take default values (`0`)?

Can we confirm this with a unit test?

cc @Kourin1996

File: contracts/Staking.sol [hunk 6]
```diff
@@ -7,14 +7,15 @@ contract Staking {
...
-    uint32 public constant MinimumRequiredNumValidators = 4;
 
     // Properties
     address[] public _validators;
     mapping(address => bool) _addressToIsValidator;
     mapping(address => uint256) _addressToStakedAmount;
     mapping(address => uint256) _addressToValidatorIndex;
     uint256 _stakedAmount;
+    uint32 _maximumNumValidators;
+    uint32 _minimumNumValidators;
```
Comment (zivkovicmilos) [inline contracts/Staking.sol:18]: You've made these values private, but haven't provided adequate getter methods.

Please consider providing them to improve the UX 🙏 

cc @Kourin1996

**Review by zivkovicmilos — APPROVED** [review]

File: contracts/Staking.sol (hunk 5 above)
Comment (PtitLuca) [reply contracts/Staking.sol:39]: You cannot deploy a contract without providing the constructor values. You can verify this by trying not to provide them in the tests.
They're here for testing purposes only, as we don't use the constructor in the predeploy process.

Comment (zivkovicmilos) [reply contracts/Staking.sol:39]: What's the default value?

Comment (PtitLuca) [reply contracts/Staking.sol:39]: Yes, in solidity the default uint32's value is zero.

File: contracts/Staking.sol (hunk 6 above)
Comment (PtitLuca) [reply contracts/Staking.sol:18]: Added in https://github.com/0xPolygon/staking-contracts/pull/10/commits/7cca3f7a5e005edabca976a999dd89e013c56ddf

File: contracts/Staking.sol (hunk 5 above)
Comment (Kourin1996) [reply contracts/Staking.sol:39]: Firstly, they are contract fields and must be initialized regardless of the difference between normal deploy and pre-deploy. The deployed bytecode may not have initialization of these fields.

Whan you deploy contract in normal way, you can't deploy contract without required arguments. Then you don't have to care about this.

@zivkovicmilos @PtitLuca

Comment (Kourin1996) [inline contracts/Staking.sol:39]: By the way, add checking not if _minimumNumValidators exceeds _maximumNumValidators in constructor is user-friendly

**Review by Kourin1996 — APPROVED** [review]

File: contracts/Staking.sol (hunk 5 above)
Comment (lazartravica) [reply contracts/Staking.sol:39]: @Aleksao998 you have fixed this in the polygon-edge pr correct? @zivkovicmilos please review

Comment (Aleksao998) [reply contracts/Staking.sol:39]: @lazartravica As @Kourin1996 said when deploying a contract usually these attributes are required. 

When we are pre-deploying with deployed bytecode, we have to set these fields manually using a storage map. We have an e2e test that checks does our default boundaries work. I have tried setting these values as

File: contracts/Staking.sol [hunk 7]
```diff
@@ -67,26 +85,27 @@ contract Staking {
...
 
         emit Staked(msg.sender, msg.value);
     }
 
     function _unstake() private {
         require(
-            _validators.length > MINIMUM_REQUIRED_NUM_VALIDATORS,
-            "Validators can't be less than MINIMUM_REQUIRED_NUM_VALIDATORS"
+            _validators.length > _minimumNumValidators,
+            "Validators can't be less than MinimumRequiredNumValidators"
```
Comment (zivkovicmilos) [inline contracts/Staking.sol:108]: Please fix this error message

File: scripts/deploy.ts [hunk 8]
```diff
@@ -1,15 +1,18 @@
 import {ethers} from "hardhat";
 
 async function main() {
+  const MinValidatorCount = 1;
+  //Max uint32
```
Comment (zivkovicmilos) [inline scripts/deploy.ts:5]: Nitpick, please add a space between `//` and the rest of the comment

File: scripts/deploy.ts [hunk 9]
```diff
@@ -1,15 +1,18 @@
...
 
   console.log("Deploying contracts with the account:", deployer.address);
   console.log("Account balance:", (await deployer.getBalance()).toString());
 
   const StakingContractFactory = await ethers.getContractFactory("Staking");
-  const stakingContract = await StakingContractFactory.deploy({
-    gasLimit: '0x100000'
-  });
+
+  const stakingContract = await StakingContractFactory.deploy(MinValidatorCount, MaxValidatorCount);
```
Comment (zivkovicmilos) [inline scripts/deploy.ts:15]: Why not have this values settable from the .env and only use defaults if they're not provided?

File: test/Staking.ts [hunk 10]
```diff
@@ -4,14 +4,15 @@ import { BigNumber } from "ethers";
...
+
 describe("Staking contract", function () {
   let accounts: SignerWithAddress[];
   let contract: Staking;
   beforeEach(async () => {
     accounts = await ethers.getSigners();
 
     const contractFactory = await ethers.getContractFactory("Staking");
-    contract = (await contractFactory.deploy()) as Staking;
+    contract = (await contractFactory.deploy(4, 6)) as Staking;
```
Comment (zivkovicmilos) [inline test/Staking.ts:15]: Please extract these magic numbers so they're explicit in the test

File: test/Staking.ts [hunk 11]
```diff
@@ -133,15 +157,15 @@ describe("Staking contract", function () {
       ).to.be.revertedWith("Only staker can call function");
     });
 
-    it("should failed if current number of validators equals to MINIMUM_REQUIRED_NUM_VALIDATORS", async () => {
+    it("should failed if current number of validators equals to MinimumRequiredNumValidators", async () => {
```
Comment (zivkovicmilos) [inline test/Staking.ts:160]: This test case name should be more user friendly, not reference literal variable names

File: test/Staking.ts [hunk 12]
```diff
@@ -199,5 +223,22 @@ describe("Staking contract", function () {
           accounts[3].address,
         ]);
     });
+
+    it("should be able to accept new validator after last one has unstaked", async () => {
```
Comment (zivkovicmilos) [inline test/Staking.ts:227]: Typo, `should be able to accept a new......after the last one...`

File: test/Staking.ts [hunk 13]
```diff
@@ -199,5 +223,22 @@ describe("Staking contract", function () {
...
+          .slice(0, 6)
+          .map((account) => {
+            stake(account, value);
+          })
+      );
+
+      // Account #0 unstakes
+      await contract.connect(accounts[0]).unstake();
+
+      // Account #6 should be able to become validator
```
Comment (zivkovicmilos) [inline test/Staking.ts:240]: Typo here, `should be able to become a validator`

**Review by zivkovicmilos — APPROVED** [review]
Overall looks good, I've left a few comments that need to be resolved 🚀

File: contracts/Staking.sol (hunk 7 above)
Comment (Aleksao998) [reply contracts/Staking.sol]: Fixed 115588a7e5c6265e64aa2ce6755bf45a100913b9

File: scripts/deploy.ts (hunk 8 above)
Comment (Aleksao998) [reply scripts/deploy.ts]: Fixed 115588a7e5c6265e64aa2ce6755bf45a100913b9

File: scripts/deploy.ts (hunk 9 above)
Comment (Aleksao998) [reply scripts/deploy.ts]: Fixed 115588a7e5c6265e64aa2ce6755bf45a100913b9

File: test/Staking.ts (hunk 10 above)
Comment (Aleksao998) [reply test/Staking.ts]: Fixed 115588a7e5c6265e64aa2ce6755bf45a100913b9

File: test/Staking.ts (hunk 11 above)
Comment (Aleksao998) [reply test/Staking.ts]: Fixed 115588a7e5c6265e64aa2ce6755bf45a100913b9

File: test/Staking.ts (hunk 12 above)
Comment (Aleksao998) [reply test/Staking.ts]: Fixed 115588a7e5c6265e64aa2ce6755bf45a100913b9

File: test/Staking.ts (hunk 13 above)
Comment (Aleksao998) [reply test/Staking.ts]: Fixed 115588a7e5c6265e64aa2ce6755bf45a100913b9

File: .env.example [hunk 14]
```diff
@@ -1,3 +1,5 @@
 JSONRPC_URL=http://localhost:10002
 PRIVATE_KEYS=0x
 STAKING_CONTRACT_ADDRESS=0x
+MAX_VALIDATOR_COUNT=1
+MIN_VALIDATOR_COUNT=4
```
Comment (zivkovicmilos) [inline .env.example:5]: These should be empty, so 
```bash
MAX_VALIDATOR_COUNT=
MIN_VALIDATOR_COUNT=
```

**Review by zivkovicmilos — APPROVED** [review]
Looks good, left a minor nitpick 💯

File: .env.example (hunk 14 above)
Comment (Aleksao998) [reply .env.example]: Fixed 60f2db0ca9d9ddad06ae03b8b8f8cfa9e346dcd9

File: test/ContractDeployment.ts [hunk 15]
```diff
@@ -0,0 +1,11 @@
...
+import {expect} from "chai";
+
+describe("Staking contract deployment", async () => {
+    it("Min validators should not be greater than max validators number", async function () {
+        const contractFactory = await ethers.getContractFactory("Staking");
+        await expect(contractFactory.deploy(7, 6)).to.be.revertedWith(
+            "Min validators num can not be greater than max num of validators"
+        );
+    })
+})
```
Comment (Kourin1996) [inline test/ContractDeployment.ts:11]: Why new file?

**Review by Kourin1996 — CHANGES_REQUESTED** [review]

File: test/ContractDeployment.ts (hunk 15 above)
Comment (Aleksao998) [reply test/ContractDeployment.ts:11]: No particular reason, I think it is clearer.
In test/ContractDeployment.ts we can write all tests for deployment (if we add any more conditions)
in test/Staking.ts we can test all edge cases for staking/unstaking.

If you do not agree I do not agree we can merge them, but I would not like to have 1 test file for the whole repo.

Comment (Kourin1996) [reply test/ContractDeployment.ts:11]: I think we should separate by Contract file (implementation). It is possible to merge test file?

Comment (Aleksao998) [reply test/ContractDeployment.ts]: Yes of course, fixed 90257b0ef518f9c4e5ebd09bd6999a842b1dc26f

**Review by Kourin1996 — APPROVED** [review]

File: test/staking.ts [hunk 16]
```diff
@@ -48,6 +95,45 @@ describe("Staking contract", function () {
...
+            accounts[1].address,
+            accounts[2].address,
+            accounts[3].address,
+            accounts[4].address,
+            accounts[6].address,
+          ]);
+
+      expect(await contract.lowestValidator()).to.eq(accounts[5].address);
+    });
   });
```
Comment (GitesParti343) [inline test/staking.ts]: ii

File: contracts/staking.sol [hunk 17]
```diff
@@ -116,4 +136,27 @@ contract Staking {
...
+            if (_addressToStakedAmount[_validators[i]] < min) {
+                min = _addressToStakedAmount[_validators[i]];
+                _lowestValidator = _validators[i];
+            }
+        }
+        _minimumStakedAmountByValidator = min;
+    }
+
+    function _appendToValidatorSet(address newValidator) private {
+        _addressToIsValidator[newValidator] = true;
```
Comment (GitesParti343) [inline contracts/staking.sol]: GitesParti343

File: test/staking.ts [hunk 18]
```diff
@@ -20,16 +20,63 @@ describe("Staking contract", function () {
     expect(await contract.stakedAmount()).to.eq(0);
   });
 
+  it("minimum staked amount should be default on deployed", async () => {
+    expect(await contract.minimumStakedAmountByValidator()).to.eq(0);
+  });
+
```
Comment (GitesParti343) [inline test/staking.ts]: ii

**Review by GitesParti343 — COMMENTED** [review]
ii
